src/Npgsql/NpgsqlConnectionStringBuilder.cs

```diff
@@ -1724,6 +1744,139 @@ protected override void GetProperties(Hashtable propertyDescriptors)
+    }
+
+    static NpgsqlConnectionStringBuilder? LoadConfigurationFromPgServiceFile(NpgsqlConnectionStringBuilder? source, string? serviceName = null)
+    { 
+        var service = serviceName ?? source?.Service ?? PostgresEnvironment.Service;
+        if (service is null)
+        {
+            return null;
+        }
+
+        if (source is not null)
+        {
+            source.Service = service;
+        }
+
+        return LoadConfigurationFromIniFile(PostgresEnvironment.UserServiceFile)
+            ?? LoadConfigurationFromIniFile(PostgresEnvironment.SystemServiceFile);
+
+        NpgsqlConnectionStringBuilder? LoadConfigurationFromIniFile(string? filePath)
+        {
+            if (filePath is null || !File.Exists(filePath))
+            {
+                return null;
+            }
+
+            var settings = ReadIniFile(filePath);
+            if (settings is null)
+            {
+                return null;
+            }
+
+            var builder = source ?? new NpgsqlConnectionStringBuilder();
+            foreach (var kv in settings)
+            {
+                if (!builder.Keys.Contains(kv.Key))
+                {
+                    builder[kv.Key] = kv.Value;
+                }
+            }
+
+            return builder;
+        }
+
+        Dictionary<string, string>? ReadIniFile(string filePath)
+        {
+            var lines = File.ReadAllLines(filePath);
+
+            Dictionary<string, string>? settings = default;
+            foreach (var rawLine in lines)
+            {
+                var line = rawLine.Trim();
+
+                // Ignore blank lines
+                if (string.IsNullOrWhiteSpace(line))
+                {
+                    continue;
+                }
+
+                // Ignore comments
+                if (line[0] is ';' or '#' or '/')
+                {
+                    continue;
+                }
+
+                // [Section:header]
+                if (line[0] == '[' && line[line.Length - 1] == ']')
+                {
+                    // All settings for the specific service has been retrieved
+                    if (settings is not null)
+                    {
+                        break;
+                    }
+
+                    // remove the brackets
+                    var sectionPrefix = line.Substring(1, line.Length - 2).Trim();
+                    // Check whether it is the specified service
+                    if (sectionPrefix == service)
+                    {
+                        settings = new Dictionary<string, string>();
+                    }
+
+                    continue;
+                }
+
+                // Skip lines if not in the section for specified service
+                if (settings is null)
+                {
+                    continue;
+                }
+
+                // key = value OR "value"
+                var separator = line.IndexOf('=');
+                if (separator < 0)
```

Comment:
Probably should also handle 0 (then `=` is the first character)

Reply:
Changed.